install-manager/src/main/java/cm/aptoide/pt/install_manager/RealApp.kt

```diff
@@ -0,0 +1,88 @@
+package cm.aptoide.pt.install_manager
+
+import cm.aptoide.pt.install_manager.dto.AppInfo
+import cm.aptoide.pt.install_manager.dto.InstallPackageInfo
+import cm.aptoide.pt.install_manager.dto.Version
+import cm.aptoide.pt.install_manager.repository.AppInfoRepository
+import kotlinx.coroutines.withContext
+import kotlin.coroutines.CoroutineContext
+
+internal class RealApp<D> internal constructor(
+  override val packageName: String,
+  private var isKnown: Boolean = false,
+  private var _installedVersion: Version? = null,
+  private var _details: D? = null,
+  private val taskFactory: Task.Factory,
+  private val appInfoRepository: AppInfoRepository<D>,
+  private val context: CoroutineContext,
+) : App<D> {
+
+  override val installedVersionName get() = _installedVersion?.versionName
+
+  override val installedVersionCode get() = _installedVersion?.versionCode
+
+  override val details get() = _details
+
+  override suspend fun getTask(): Task? = withContext(context) {
+    taskFactory.getTask(packageName)
+  }
+
+  override suspend fun setDetails(details: D) = withContext(context) {
+    if (isKnown) {
+      appInfoRepository.save(AppInfo(packageName, _installedVersion, details))
+    }
+    _details = details
+  }
+
+  override suspend fun install(installPackageInfo: InstallPackageInfo): Task {
+    return when {
+      getTask() != null -> {
+        throw IllegalStateException("another task is already queued")
+      }
+      installPackageInfo.version.versionCode == _installedVersion?.versionCode -> {
+        throw IllegalArgumentException("This version is already installed")
+      }
+      installPackageInfo.version.versionCode < (_installedVersion?.versionCode
+        ?: Int.MIN_VALUE) -> {
+        throw IllegalArgumentException("Newer version is installed")
+      }
```

Comment:
I think it could be achieved on client side. Something like this for instance if we create a custom exception:
```
try {
    installManager.getApp("package").install(installInfo)
} catch() {
    installManager.getApp("package").uninstall().collect()
    installManager.getApp("package").install(installInfo)
}
```

Or simply calling these both if it is known in advance that it will be a downgrade
```
    installManager.getApp("package").uninstall().collect()
    installManager.getApp("package").install(installInfo)
```

In both cases end user should be aware about loosing all app data after that...

If we still need a dedicated method for that with it's own status-progress flow, I can add it. Need some business logic details about it, though 🤔